util/helper-cairo-ansi.cc

```diff
@@ -26,9 +26,102 @@
 
 #include "helper-cairo-ansi.hh"
 #include "options.hh"
-
 #include "ansi-print.hh"
 
+#ifdef HAVE_CHAFA
+# include <chafa.h>
+
+/* Similar to ansi-print.cc */
+# define CELL_W 8
+# define CELL_H (2 * CELL_W)
+
+static void
+chafa_print_image_rgb24 (const void *data, int width, int height, int stride)
+{
+  ChafaTermInfo *term_info;
+  ChafaSymbolMap *symbol_map;
+  ChafaCanvasConfig *config;
+  ChafaCanvas *canvas;
+  GString *gs;
+  unsigned int cols = (width +  CELL_W - 1) / CELL_W;
+  unsigned int rows = (height + CELL_H - 1) / CELL_H;
+  gchar **environ;
+  ChafaCanvasMode mode;
+  ChafaPixelMode pixel_mode;
+
+  /* Adapt to terminal; use sixels if available, and fall back to symbols
+   * with as many colors as are supported */
+
+  environ = g_get_environ ();
+  term_info = chafa_term_db_detect (chafa_term_db_get_default (),
+                                    environ);
+
+  pixel_mode = CHAFA_PIXEL_MODE_SYMBOLS;
+
+  if (chafa_term_info_have_seq (term_info, CHAFA_TERM_SEQ_BEGIN_SIXELS))
+  {
+    pixel_mode = CHAFA_PIXEL_MODE_SIXELS;
+    mode = CHAFA_CANVAS_MODE_TRUECOLOR;
+  }
+  else if (chafa_term_info_have_seq (term_info, CHAFA_TERM_SEQ_SET_COLOR_FGBG_DIRECT))
+    mode = CHAFA_CANVAS_MODE_TRUECOLOR;
+  else if (chafa_term_info_have_seq (term_info, CHAFA_TERM_SEQ_SET_COLOR_FGBG_256))
+    mode = CHAFA_CANVAS_MODE_INDEXED_240;
+  else if (chafa_term_info_have_seq (term_info, CHAFA_TERM_SEQ_SET_COLOR_FGBG_16))
+    mode = CHAFA_CANVAS_MODE_INDEXED_16;
+  else if (chafa_term_info_have_seq (term_info, CHAFA_TERM_SEQ_INVERT_COLORS))
+    mode = CHAFA_CANVAS_MODE_FGBG_BGFG;
+  else
+    mode = CHAFA_CANVAS_MODE_FGBG;
+
+  /* Create the configuration */
+
+  symbol_map = chafa_symbol_map_new ();
+  chafa_symbol_map_add_by_tags (symbol_map,
+                                (ChafaSymbolTags) (CHAFA_SYMBOL_TAG_BLOCK
+                                                   | CHAFA_SYMBOL_TAG_SPACE));
+
+  config = chafa_canvas_config_new ();
+  chafa_canvas_config_set_canvas_mode (config, mode);
+  chafa_canvas_config_set_pixel_mode (config, pixel_mode);
+  chafa_canvas_config_set_cell_geometry (config, 10, 20);
```

Comment:
Curious. How are these numbers used?

Reply:
The cell geometry is the pixel size of a single character cell, and the canvas geometry is the size of the canvas in character cells. The cell size is only used in sixel mode: Let's say you have a canvas that's 10 cells wide, and we supply that character cells are 10 pixels wide. Chafa will then produce a sixel image that's 100 pixels wide, scaling the input image if necessary.

I decided to use 10x20 because that's the fixed cell size of VT240 and VT340 terminals (and possibly Windows Terminal too when/if it gets sixel support).

This will lead to sixel output being a bit bigger than expected on modern Linux terminals (character cells tend to be bigger than 10x20, but the aspect is usually about the same). It would be possible to get the cell size with the  `TIOCGWINSZ` ioctl (.ws_xpixel and .ws_ypixel fields divided by .ws_col and .ws_row). The `chafa` command-line tool does this, but I avoided it in this patch because it seemed minor and might make Harfbuzz less portable.

See https://github.com/hpjansson/chafa/blob/2af7edec1adfdf1825cfb0dba46e8aaed7504918/tools/chafa/chafa.c#L726-L753 for how I did it in `chafa`. I can add something similar to `hb-view` if you think it's worth it.

Also see [here](https://gitlab.gnome.org/GNOME/vte/-/issues/312) for a long and somewhat opinionated discussion of sixel/cell dimensions arcana :)